### PR TITLE
Bug 1834311: Fix trusted ca bundle and hash reconciliation

### PR DESCRIPTION
**Pull request by periklis**
This PR addresses several issues on reconciliation of the trusted CA-Bundle config map as well the hash annotation on the Kibana deployment.

To address:
- https://bugzilla.redhat.com/show_bug.cgi?id=1834311
- https://bugzilla.redhat.com/show_bug.cgi?id=1833273

Needs backport to: release-4.4

### CODE REVIEW

File: pkg/k8shandler/kibana/configmap.go [hunk 1]
```diff
@@ -52,12 +44,10 @@ func (clusterRequest *KibanaRequest) createOrUpdateConfigMap(configMap *core.Con
 				return fmt.Errorf("Failed to get %v configmap for %q: %v", configMap.Name, clusterRequest.cluster.Name, err)
 			}
 
-			if checkData {
```
Comment (jcantrill) [inline pkg/k8shandler/kibana/configmap.go:55]: This change appears valid but what is "checkData" ?

Comment (periklis) [reply pkg/k8shandler/kibana/configmap.go:55]: Check L36 above. We used to check data for all config maps except for the TrustedCABundle. The trusted CA bundle logic didn't belong here in the first place. Thus, I dropped the extra check

File: pkg/controller/trustedcabundle/controller.go [hunk 2]
```diff
@@ -64,23 +64,21 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
...
-// collector and the visualization resources.
+// Reconcile reads that state of the trusted CA bundle configmap objects for Kibana resource.
 // When the user configured and/or system certs are updated, the pods are triggered to restart.
 func (r *ReconcileTrustedCABundle) Reconcile(request reconcile.Request) (reconcile.Result, error) {
-	if request.Name == constants.KibanaTrustedCAName {
-		if err := kibanahandler.ReconcileKibanaInstance(request, r.client); err != nil {
-			// Failed to reconcile - requeuing.
-			return reconcileResult, err
-		}
+	if err := kibana.ReconcileKibanaInstance(request, r.client); err != nil {
```
Comment (ewolinetz) [inline pkg/controller/trustedcabundle/controller.go:79]: so the reason we had the check before was because we only wanted to reconcile Kibana when our expected trustbundle was updated.... if this is causing it to not go through then something else may be the issue...

File: pkg/k8shandler/elasticsearch.go [hunk 3]
```diff
@@ -475,8 +475,9 @@ func (elasticsearchRequest *ElasticsearchRequest) AddAliasForOldIndices() bool {
 		}
 
 		if payload.ResponseBody[index] != nil {
-			indexBody := payload.ResponseBody[index].(map[string]interface{})
-			if indexBody["aliases"] != nil {
+			indexRawBody, okIndex := payload.ResponseBody[index]
```
Comment (ewolinetz) [inline pkg/k8shandler/elasticsearch.go:478]: nit but this seems out of scope for this bug fix... can we open up another bz to track this work?

File: pkg/k8shandler/kibana/kibana_test.go [hunk 4]
```diff
@@ -2,19 +2,18 @@ package kibana
...
 	"context"
-	"fmt"
 
 	. "github.com/onsi/ginkgo"
 	. "github.com/onsi/gomega"
 	configv1 "github.com/openshift/api/config/v1"
 	consolev1 "github.com/openshift/api/console/v1"
 	routev1 "github.com/openshift/api/route/v1"
-	kibana "github.com/openshift/elasticsearch-operator/pkg/apis/logging/v1"
+	kibanav1 "github.com/openshift/elasticsearch-operator/pkg/apis/logging/v1"
```
Comment (ewolinetz) [inline pkg/k8shandler/kibana/kibana_test.go:11]: nit but this would make more sense as `loggingv1` wouldn't it? since we don't have a `kibana` api endpoint...

Comment (periklis) [reply pkg/k8shandler/kibana/kibana_test.go:11]: Indeed!

File: pkg/k8shandler/elasticsearch.go (hunk 3 above)
Comment (periklis) [reply pkg/k8shandler/elasticsearch.go:478]: Prove me wrong, but i see regurarly the operator panic'ing on e2e.

File: pkg/controller/trustedcabundle/controller.go (hunk 2 above)
Comment (periklis) [reply pkg/controller/trustedcabundle/controller.go:79]: I believe this was an obsolete double check, because already handle this checks in https://github.com/openshift/elasticsearch-operator/pull/351/files#diff-d77a5e8d053611814893162d4e4d0b18R52

File: pkg/k8shandler/kibana/trustedcabundle.go [hunk 5]
```diff
@@ -27,11 +29,26 @@ func (clusterRequest *KibanaRequest) createOrUpdateTrustedCABundleConfigMap(conf
...
-	err := clusterRequest.CreateOrUpdateTrustedCaBundleConfigMap(configMap)
-	return err
+	err := clusterRequest.Create(configMap)
+	if err != nil {
+		if !errors.IsAlreadyExists(err) {
+			return nil, fmt.Errorf("failed to create trusted CA bundle config map %q for %q: %s", name, clusterRequest.cluster.Name, err)
+		}
+
+		// Get the existing config map which may include an injected CA bundle
+		if err = clusterRequest.Get(configMap.Name, configMap); err != nil {
```
Comment (ewolinetz) [inline pkg/k8shandler/kibana/trustedcabundle.go:39]: nit, but since we aren't doing anything with the CM other than ensuring it isn't nil (which should happen once we set it up for creating). not sure we need to do the extra api call ... we are already doing a `get` to build the hash in the `getAnnotations` function

**Comment by ewolinetz** [comment]
@periklis couple nits otherwise :+1:

File: pkg/k8shandler/elasticsearch.go (hunk 3 above)
Comment (periklis) [reply pkg/k8shandler/elasticsearch.go:478]: I'll remove and open a separate changeset.

File: pkg/k8shandler/kibana/trustedcabundle.go (hunk 5 above)
Comment (periklis) [reply pkg/k8shandler/kibana/trustedcabundle.go:39]: It's not valid to drop this `Get` call actually because the refreshed config map is used in building the pod spec. In the absence of a valid CA bundle the pod factory function does not add the volume/volumemount to the spec.

File: pkg/controller/trustedcabundle/controller.go (hunk 2 above)
Comment (ewolinetz) [reply pkg/controller/trustedcabundle/controller.go:79]: yup, i realized this after looking at your CLO change. my mistake :)

File: pkg/k8shandler/kibana/trustedcabundle.go (hunk 5 above)
Comment (ewolinetz) [reply pkg/k8shandler/kibana/trustedcabundle.go:39]: ah, I didn't realize that `hasTrustedCABundle` ensures it's been populated
https://github.com/openshift/elasticsearch-operator/blob/master/pkg/k8shandler/kibana/trustedcabundle.go#L34-L44
